scripts/promote-catalog.py

```diff
@@ -0,0 +1,486 @@
+    """
+    Main function
+    """
+    parser = setup_argparser()
+    args = parser.parse_args()
+    LOG.setLevel(logging.INFO)
+    if args.verbose:
+        LOG.setLevel(logging.DEBUG)
+    LOG.addHandler(logging.StreamHandler())
+
+    LOG.info("Starting promotion process")
+
+    local_repo_name = os.path.basename(args.local_repo.strip("/"))
+    try:
+        organization = ORGANIZATIONS.get(local_repo_name, {})["organization"]
+    except KeyError:
+        LOG.error("Repository %s not supported", local_repo_name)
+        return
+
+    # check repo status
+    # all new branches will be based on actual branch
+    # all previous changes will be lost
+    git_repo = GitRepo(args.local_repo)
+    base_branch = str(git_repo.active_branch)
+    if not repo_status_clean(git_repo):
+        LOG.warning(
+            "The repo is either dirty or not on main branch, "
+            "please check its status. "
+            "If you decide to continue, the uncommited changes "
+            "and untracked files will be lost."
+        )
+        its_ok = input("Do you want to continue? [y/N]: ") or "n"
+        if its_ok.lower() != "y":
+            LOG.info("Exiting")
+            return
+
+    create_or_clear_branch_and_checkout(git_repo)
+
+    # configure repo with the token
+    access_token = os.environ.get("GITHUB_TOKEN")
+    gh_repository = ORGANIZATIONS[local_repo_name]["gh_repository"]
+    remote_url_with_token = f"https://{access_token}@github.com/{gh_repository}.git"
+    git_repo.remotes.origin.set_url(remote_url_with_token)
+
+    # gather ocp_version related info
+    target_version = args.target_version
+    pyxis_url = args.pyxis_url
+    ocp_versions = ocp_version_info(None, pyxis_url, organization)
+    # `indices` are sorted by version in descending order
+    valid_versions = [v.get("ocp_version") for v in ocp_versions["indices"]]
+    # check if target version index exists
+    if target_version and target_version not in valid_versions:
+        LOG.error(
+            "Target version %s not in valid versions %s", target_version, valid_versions
+        )
+        return
+
+    # default to max version index
+    if not target_version:
+        target_version = ocp_versions["max_version_index"].get("ocp_version")
+
+    # process promotion
+    LOG.info("Promotion to version: %s", target_version)
+    pr_info, branch_suffix = promote_catalog(
+        args.local_repo, git_repo, valid_versions, target_version
+    )
+    LOG.info("Suffix for all branches created and pushed: %s", branch_suffix)
+
+    # dry run only pushes branches without creating PRs
+    # this can be used to prepare PR branches and then create PRs manually
+    # to avoid multiple pipelines starting at once
+    if pr_info and not args.dry_run:
+        for push_data in pr_info:
+            create_pr(
+                base_branch,
+                push_data["head"],
+                push_data["title"],
+                local_repo_name,
+            )
```

Comment:
Just an idea: since this could potentially create a lot of PRs at once (especially in the community repo where I expect more people will choose the "always" promotion strategy), what about adding a label to the PRs created by this automation to make it easier to filter and search for PRs in the web ui?

Reply:
I'm still not sure how we would handle the bulk of PRs, therefore I implemented the dry-run to create only branches and then we can manually create PRs - branches are identified by the suffix. I can add the label, but unfortunately it cannot be part of the PR creation api request and the label must be created separately and then applied in an additional API call. If you think it is worth the effort and the branch suffix is not enough, then I can add it?